avfilter: Add tonemap vaapi filter for H2S

sunxinpeng opened the pull request:
It performs HDR(High Dynamic Range) to SDR(Software Defination Radio) conversion
with tone-mapping. It supports HDR10 only as input temporarily.

H2S: P010 -> NV12

sign-off-by: Xinpeng Sun <xinpeng.sun@intel.com>
sign-off-by: Zachary Zhou <zachary.zhou@intel.com>

sunxinpeng commented on libavfilter/vf_tonemap_vaapi.c:
```diff
@@ -0,0 +1,270 @@
+    if (err < 0)
+        goto fail;
+
+    av_frame_free(&input_frame);
```
It's a common behavior to free the input_frame buffer when the filter call is over.

sunxinpeng commented on libavfilter/vf_tonemap_vaapi.c:
```diff
@@ -0,0 +1,270 @@
+           input_frame->width, input_frame->height, input_frame->pts);
+
+    if (vpp_ctx->va_context == VA_INVALID_ID)
+        return AVERROR(EINVAL);
```
It's no need to release the input buffer here because it will be used by the upper layer call.

xhaihao replied:
If so, why is the input buffer is released below for other errors?

sunxinpeng commented on libavfilter/vf_tonemap_vaapi.c:
```diff
@@ -0,0 +1,270 @@
+
+static int tonemap_vaapi_vpp_query_formats(AVFilterContext *avctx)
+{
+    return ff_vaapi_vpp_query_formats(avctx);
```
If ff_formats_ref return error value, it doesn't need to use ff_formats_unref to free the memory.

xhaihao replied:
If so, when will the memory be released?

xhaihao commented:
SDR stands for Standard Dynamic Range, see https://en.wikipedia.org/wiki/Standard-dynamic-range_video

> It supports HDR10 only as input **temporarily**

Does it mean this is not a stable version?

xhaihao commented on doc/filters.texi:
```diff
@@ -20754,6 +20754,36 @@ Convert HDR(PQ/HLG) video to bt2020-transfer-characteristic p010 format using li
 
+@section tonemap_vappi
+
+Perform HDR(High Dynamic Range) to SDR(Software Defination Radio) conversion with tone-mapping.
```
SDR stands for Standard-dynamic-range

xhaihao commented on doc/filters.texi:
```diff
@@ -20754,6 +20754,36 @@ Convert HDR(PQ/HLG) video to bt2020-transfer-characteristic p010 format using li
+@section tonemap_vappi
+
+Perform HDR(High Dynamic Range) to SDR(Software Defination Radio) conversion with tone-mapping.
+It maps the dynamic range of HDR10 content to the SDR display panel.
```
is the result used for display only ?  I don't think H2S has this limitation.

xhaihao commented on doc/filters.texi:
```diff
@@ -20754,6 +20754,36 @@ Convert HDR(PQ/HLG) video to bt2020-transfer-characteristic p010 format using li
+
+@itemize
+@item
+Convert HDR video to SDR video from p010 format to argb format.
```
Conflict with the statement for H2S.

sunxinpeng reviewed (commented):
> SDR stands for Standard Dynamic Range, see https://en.wikipedia.org/wiki/Standard-dynamic-range_video
> 
> > It supports HDR10 only as input **temporarily**
> 
> Does it mean this is not a stable version?

Media driver only supports HDR10 by now, but it seems that other HDR standards(like HLG) are in the to do list.

xhaihao commented on libavfilter/vf_tonemap_vaapi.c:
```diff
@@ -0,0 +1,270 @@
+    case HDR_VAAPI_H2S:
+        params.output_color_standard = VAProcColorStandardBT709;
+	    params.output_color_properties.colour_primaries = AVCOL_PRI_BT709;
+	    params.output_color_properties.transfer_characteristics = AVCOL_TRC_BT709;
```
Please use the same indentation style.

fulinjie commented:
IMHO, this PR is not fully checked(since it's only  checked by bionic/pr)

xenial/pr/compile would check whether this PR has some compatible issued with early version libva.

zhoujd commented:
Please make you have all 5 checks passed, this PR failed on https://media-ci.ostc.intel.com/build/348719/overview

ruiling commented on doc/filters.texi:
```diff
@@ -20769,6 +20769,36 @@ Convert HDR(PQ/HLG) video to bt2020-transfer-characteristic p010 format using li
+
+@table @option
+@item type
+Specify the tone-mapping operator to be used.
```
I would like you to remove the 'type' option. it is easy to derive it from the color_transfer. if the output is smpte2084 or HLG, you can easily say this is H2H, otherwise H2S. I still think tone-mapping to 10bit SDR is important, please teach me if I am wrong. so please help discuss with media-driver developer to add 10bit SDR support.

sunxinpeng replied:
I agree with you. But I think it is more clear for user to use this filter if keeping this option. How about keeping the 'type' option in this version?

ruiling replied:
I didn't see 'more clear'. this is redundant with output transfer function. take a look at other tone-mapping filters in FFmpeg.

sunxinpeng replied:
Fixed in the new version.

ruiling commented on libavfilter/vaapi_vpp.c:
```diff
@@ -257,6 +257,11 @@ static const VAAPIColourProperties vaapi_colour_standard_map[] = {
+//#if VA_CHECK_VERSION(2, 3, 0)
+//    { VAProcColorStandardExplicit,    9,  16, AVCOL_SPC_BT2020_NCL},
+//#endif
+
```
why these lines was commented?

sunxinpeng replied:
> why these lines was commented?

"If explicit values are provided then you don't need it in this map."
----from Mark T.

ruiling commented on libavfilter/vf_tonemap_vaapi.c:
```diff
@@ -0,0 +1,454 @@
+        output_frame->color_trc = ctx->colour_transfer;
+    } else {
+        switch (vpp_ctx->output_format) {
+        case AV_PIX_FMT_NV12:
```
i don't think you need to set the output color_trc based on the output_format. You can simply set to BT709 if user does not give one.

ruiling commented on libavfilter/vf_tonemap_vaapi.c:
```diff
@@ -0,0 +1,454 @@
+        switch (vpp_ctx->output_format) {
+        case AV_PIX_FMT_NV12:
+            output_frame->colorspace = AVCOL_SPC_BT709;
+            break;
```
i don't think you need to change the output frame's colorspace, if user does not give one. just use the input's colorspace, av_frame_copy_props() already does this.

ruiling commented on libavfilter/vf_tonemap_vaapi.c:
```diff
@@ -0,0 +1,454 @@
+            break;
+        default:
+            av_assert0(0);
+        }
```
do't change the output frame's color-primaries if user does not set one.

xhaihao commented on doc/filters.texi:
```diff
@@ -20819,6 +20819,8 @@ Convert HDR(PQ/HLG) video to bt2020-transfer-characteristic p010 format using li
 @end itemize
 
+
+
```
Please remove the above new lines.

ruiling commented on doc/filters.texi:
```diff
@@ -20879,6 +20881,105 @@ Apply a strong blur of both luma and chroma parameters:
+@item
+Example of running tonemap_vaapi filter with default parameters on it.
+@example
+-hwaccel vaapi -vaapi_device /dev/dri/renderD128  -hwaccel_output_format vaapi -i INPUT -vf "hwupload, tonemap_vaapi, hwdownload" OUTPUT
```
are you sure this command line really need hwupload and hwdownload??

ruiling commented on doc/filters.texi:
```diff
@@ -20879,6 +20881,105 @@ Apply a strong blur of both luma and chroma parameters:
+@end table
+
+Default is same as input.
+
```
for the primaries transfer matrix, you need to check what values are supported by vaapi.

ruiling commented on libavfilter/vf_tonemap_vaapi.c:
```diff
@@ -0,0 +1,437 @@
+    av_assert0(ctx->color_primaries == AVCOL_PRI_BT2020 ||
+               ctx->color_primaries == AVCOL_PRI_BT709 ||
+               ctx->color_primaries == AVCOL_PRI_UNSPECIFIED);
+
```
Is this really the limitation of vaapi tone-mapping?

ruiling commented on libavfilter/vf_tonemap_vaapi.c:
```diff
@@ -0,0 +1,437 @@
+
+AVFilter ff_vf_tonemap_vaapi = {
+    .name           = "tonemap_vaapi",
+    .description    = NULL_IF_CONFIG_SMALL("VAAPI VPP for tonemap"),
```
tone-mapping instead of tonemap is more suitable here.

fulinjie commented:
@sunxinpeng This still has compatible issues and would  break build for early version of libva.

Hi @zhoujd , only two checks have been run on this PR instead of 5 checks. And it hides the potential compatible issues, anything we could do to help and avoid this?

zhoujd commented:
Thanks Linjie, I triggered all five build for this PR,  this strange thing is CI server ran the builds build, but the results not show on github page, I will look into it.

fulinjie commented:
Hi @zhoujd, currently CI is using the latest libva version in xenial/pr/compile to check the compatible issue:
https://github.com/intel/libva/commit/09c8c2e20b7763fc44299172dbb16061b57f9c71

This makes no sense, would you please help to fix?

zhoujd commented:
@fulinjie, please understand current xenial/pr/compile does two versions of  libva, 
one is "Build with System VA-API " by use apt install libva-dev
another is " Build With Mainline VA-API" for master

fulinjie commented:
> @fulinjie, please understand current xenial/pr/compile does two versions of libva,
> one is "Build with System VA-API " by use apt install libva-dev
> another is " Build With Mainline VA-API" for master

Got it, thanks for the explains.

sunxinpeng committed:
avfilter: add overlay_vaapi filter

Overlay one video on the top of another.

It takes two inputs and has one output. The first input is the "main" video on which the
second input is overlaid. This filter requires same memory layout for all the inputs.

An example command to use this filter to overlay an image LOGO at the top-left corner of
the INPUT video and both inputs are yuv420p format:
FFMPEG -hwaccel vaapi -vaapi_device /dev/dri/renderD128 -hwaccel_output_format vaapi \
-i INPUT -i LOGO -filter_complex \
"[0:v]hwupload[a], [1:v]format=yuv420p, hwupload[b], [a][b]overlay_opencl, hwdownload" \
OUTPUT

Signed-off-by: Xinpeng Sun <xinpeng.sun@intel.com>
Signed-off-by: Zachary Zhou <zachary.zhou@intel.com>
